### ISSUE TITLE
Distributed tracing doesn't correlate when using System.Diagnostics.DiagnosticSource 4.7.1

### ISSUE THREAD

**Issue by NickSevens** [Azure/azure-functions-durable-extension#1615]
### Description
I've setup distributed tracing for Application Insights in my durable functions.
The DtOrchestrator and DtActivity items are correlated correctly, but my Traces (LogInformation, LogDebug, LogTrace) are not correlated to the Activities anymore.
I'm pretty sure this worked a while ago, and also in the early preview NuGet package.
The traces _are_ logged to AI, just not linked to any DtActivity or DtOrchestrator instance.

### Expected behavior
I expect to see the Traces correlated to a specific DtActivity, so I can filter them by Activity.
In this pane for example:
![image](https://user-images.githubusercontent.com/1467263/102373249-4a6f2200-3fc0-11eb-8a33-0ca3e1565a18.png)

### Actual behavior
The Trace messages are not correlated to the Activity, so I have to search for them separately to browse logs.

### Relevant source code snippets
Logging setup has been tried with and without `W3CDistributedTracing` enabled
e.g.:
```json
{
  "version": "2.0",
  "logging": {
    "logLevel": {
      "Default": "Information",
      "DurableTask": "Warning",
      "Function": "Trace",
      "Host": "Warning",
      "Microsoft.Azure.WebJobs": "Warning"
    },
    "applicationInsights": {
      "httpAutoCollectionOptions": {
        "enableW3CDistributedTracing": true
      },
      "samplingExcludedTypes": "Request",
      "samplingSettings": {
        "isEnabled": true
      }
    }
  },
  "extensions": {
    "durableTask": {
      "hubName": "WorkspacesHub",
      "tracing": {
        "distributedTracingEnabled": true,
        "distributedTracingProtocol": "W3CTraceContext"
      }
    }
  }
}
```

### Known workarounds
(None)

### App Details

- **Durable Functions extension version**: v2.4.0
- **Azure Functions runtime version**: v3
- **Programming language used**: C#

### Screenshots
As an example, these are the properties returned for a DtActivity and should-be-correlated Trace message
![image](https://user-images.githubusercontent.com/1467263/102374133-442d7580-3fc1-11eb-914d-612f9d60355b.png)
![image](https://user-images.githubusercontent.com/1467263/102374145-48599300-3fc1-11eb-9022-d5e6bf047b72.png)

**Comment by davidmrdavid** [comment]
Hi @NickSevens, thank you for bringing this to our attention!

I'm sorry this isn't working as expected, the feature is still in early development and was only recently merged into the main package release, so it's definitely possible some older features stopped working. I'll take a look at this.

In the meantime, it would help me greatly if you could provide a minimal reproducer that illustrates how you're generating these traces and that I could use to determine if your issue has been fixed. Any chance you could provide that? _Thanks!_

⚡  ⚡

**Comment by NickSevens** [comment]
Hi @davidmrdavid, I tried creating a minimal sample, but there it _does_ correlate correctly. So ... I'm now trying to figure out what's different in my app. :)
I'll post here if I find anything causing the issue at my end...

**Comment by mpaul31** [comment]
i also noticed the operation name is `<Empty>` in the portal now after the upgrade so unfortunately this feature is not mature enough for our production workloads quiet yet.

**Comment by davidmrdavid** [comment]
Hi @mpaul31, you're correct. The feature is still in public preview so the experience needs some more work before being ready for production. Just for my reference, where are you seeing this `<Empty>` operation name. Do you mind clarifying / sharing a screenshot _Thanks!_

**Comment by mpaul31** [comment]
Sure. You can see it from the performance blade:

![image](https://user-images.githubusercontent.com/206767/102413430-e437ce80-3fc2-11eb-9c6b-f63d280821b4.png)

**Comment by davidmrdavid** [comment]
@mpaul31 , thanks. I'll turn that into a separate ToDo item and prioritize it for an upcoming release of Dist. Tracing

**Comment by NickSevens** [comment]
@davidmrdavid I've narrowed it down to this:
I recently upgraded a NuGet package which has a dependency on `System.Diagnostics.DiagnosticSource`. The update caused the dependency to go from `4.6.0` to `4.7.1`. This causes the correlation to fail.

When the correlation fails, I noticed a difference in Operation ID format:
- _Working_ correlation Operation ID has a format of for example `00-dc5421953c0c76468418424c05b840ce-70ae1d9822e11c44-00`
- _Failing_ correlation Operation ID has a format of for example `a61c9fdd-4739b89b1bfe50b0`

As a reproduction, you can simply create a new project, and add the NuGet package.
`<PackageReference Include="System.Diagnostics.DiagnosticSource" Version="4.6.0" />` --> works
`<PackageReference Include="System.Diagnostics.DiagnosticSource" Version="4.7.1" />` --> fails

**Comment by davidmrdavid** [comment]
Thanks @NickSevens!. I'll need to dive into the documentation to understand why this would break our expected behavior. Thanks for providing this context, this is _really_ helpful :)

**Comment by NickSevens** [comment]
Not sure if it's the same, but https://tsuyoshiushio.medium.com/system-diagnostics-activity-update-and-distributed-tracing-in-c-f02428dcd050 explains something similar as well.

**Comment by lindgrend** [comment]
Having the same issue, also if you in the c# sample use Microsoft.Azure.WebJobs.Extensions.DurableTask Version="2.4.0" it also breaks for me. Maybe this can be used as your minimum repro?

**Comment by NickSevens** [comment]
Hi @davidmrdavid, do you have any update on the issue? Anything you can share? :)
Not that it's a blocking issue, but still curious ;)

**Comment by santo2** [comment]
Hi guys, is this being planned to fix? we were already looking forward to this a couple of months before the actual release, it would really help us find bugs in our system. Now it's really hard to correlate all activities together.

**Comment by davidmrdavid** [comment]
Hi folks! Unfortunately we've been slowed down by prioritizing [this](https://github.com/Azure/azure-functions-durable-extension/issues/1611) major bug but I'll take a different approach during our next release sprint and prioritize this instead. I think the bug I linked above is going to be solved on a much longer timeframe than I originally expected, so I can allow myself to multi-task on the issues moving forward :) 

Expect this to be worked on for our upcoming milestone. We haven't gotten to the release planning stage yet (we just had a release) but I understand we should be starting that process soon.

**Comment by santo2** [comment]
Hi @davidmrdavid any news on this?

**Comment by mauve** [comment]
Hi, any update?

**Comment by davidmrdavid** [comment]
Hi all, sorry for my delayed response, I somehow missed the notification on April 14th.

Thankfully, we do have several updates! Let me list them below, starting with some general news on Distributed Tracing support:

- We've been working on several Distributed Tracing improvements as of recent, and some of them will be made available with our next release, which is scheduled to take place in the coming days :tada: . That version will be `v2.5.0`
- For visibility, you can track the work items we've tackled so far, what's in progress, and what's still to-do in this Kanban board: https://github.com/Azure/azure-functions-durable-extension/projects/1#card-60122528
- @bachuv has been leading this effort, so she's probably the best person to reach for updates on this front.

With regards to this specific ticket, unfortunately there seem to be multiple different kinds of issues being reported here, so I'll try to address all concerns below.

(1) With respect to @NickSevens's original issue, which is about Traces not being correlated to Activities, we believe this issue should be fixed with our next release. By this we mean that, in our latest attempt to reproduce this issue, we do see activity traces being correlated. We're still working on correlating orchestration traces, but we describe a workaround to support this scenario here: https://github.com/Azure/azure-functions-durable-extension/discussions/1788#discussioncomment-671121

(2) @mpaul31 reported another issue: the operation name is `<Empty>` in the portal. Unfortunately, we haven't been able to reproduce this. Since we're nearing another release (one with various distributed tracing improvements), it would be good to validate if the operation name remains empty with that new release. If it does, we should create another issue to track it.
 
And that's all! Please let me, and @bachuv, know if there's anything more we can do to help!

**Comment by mauve** [comment]
Hi @davidmrdavid thanks for the detailed update, is v2.5.0 still planned to be released on the 28th of May (or around that date)?

**Comment by NickSevens** [comment]
Hi @davidmrdavid does this mean that using `System.Diagnostics.DiagnosticSource 4.7.*` still won't correlate correctly - without using a workaround?
Also: does the workaround described work with `ILogger`? it seems to be working for direct usage of `TelemetryClient` only?

We're beginning to notice issues updating other packages, which have their dependencies updated as well, which is starting to be blocking.

**Comment by davidmrdavid** [comment]
Hi @mauve. Yes, the update is still planned to be released **around** May 28th, most likely to be released a few days after that due to some extra validation steps we're trying for the first time. You can track our release progress here: https://github.com/Azure/azure-functions-durable-extension/issues/1852 . 

Relatedly, we always create release checklist-issues like the one above around our release dates, so looking for them is a good way to get up-to-date information on how the release is progressing :) .

@NickSevens, I'll defer those questions to @bachuv, who should have more context on these.

**Comment by bachuv** [comment]
@NickSevens, using `System.Diagnostics.DiagnosticSource 4.7.*` doesn't currently correlate correctly. We would need to do more investigation to understand what changes would need to be made to support that version. 

The workaround works with `TelemetryClient` since certain operation id properties have to be set for it to correlate correctly. #1805 tracks this issue.

Can you create a new GitHub issue and give us more information about what package updates are conflicting with Distributed Tracing? That way we can take a look at those packages and see if there's a workaround or solution. Thanks!

**Comment by NickSevens** [comment]
@bachuv I'm confused. This issue was created specifically since I had an issue when using `System.Diagnostics.DiagnosticSource 4.7.*`, which was explained as such [here](https://github.com/Azure/azure-functions-durable-extension/issues/1615#issuecomment-747077687).
I can surely create a new issue, but what would be the difference with this issue be?

cc: @davidmrdavid

**Comment by bachuv** [comment]
@NickSevens, sorry for any confusion! I was suggesting creating a new issue since you had mentioned having issues with updating other packages as well. Feel free to give us more information about those other packages here and we will take a look!

**Comment by mauve** [comment]
I just used/ran a distributed function with System.Diagnostics.DiagnosticSource 5.0.0 and Durablefunctions 2.5.0 and the traces are not being correlated.

DtOrchestrator has:
  - operation_Id: 55ed8602ae2ba74f966a6ae619318bcc
  - parent_Id: 2a86e9c67a98b54e

Activity has (this works):
  - operation_Id: 55ed8602ae2ba74f966a6ae619318bcc
  - parent_Id: 0e809ffa2f011b41

However my `log.logInformation()` messages have (logging from inside the above activity):
  - operation_Id: 16807b749f648445b61ccd1a8c44d1a5
  - parent_Id: 15b10a02d2b00640

**Comment by davidmrdavid** [comment]
@mauve, could you please open up a new ticket describing this issue? Since this reproduction is for a **different** DiagnosticSource version, it might be best to separate the threads :)

**Comment by NickSevens** [comment]
@bachuv @davidmrdavid is there any progress on this?

**Comment by santo2** [comment]
waiting for a long time ;-) just letting you know we would really appreciate this

**Comment by bachuv** [comment]
@NickSevens @santo2 -  We are actively working on making improvements in this area, but for now, to get traces to correlate in Activity functions, can you try following [this method from the SimpleCorrelationDemo.cs](https://github.com/Azure/azure-functions-durable-extension/blob/422eb32da7f7f5c0f75ad3170d3e32c6cca6ec9a/samples/correlation-csharp/FunctionAppCorrelation/SimpleCorrelationDemo.cs#L65) to log traces using a `TelemetryClient`? Logging traces using this method should correlate the traces with the Activity function. 

We also upgraded the System.Diagnostics.DiagnosticSource package version in [this PR](https://github.com/Azure/durabletask/pull/659) and it should be a part of our next release.

**Comment by dlebee** [comment]
I created a project to demonstrate the issue in case anyone needs it.

https://github.com/dlebee/corrolation-function-servicebus

**Comment by JakubMajsniar** [comment]
Do you have any findings about this, guys ?
I've tried to upgrade everything (AZ functions version, target framework, all packages) in my incubator project for end-to-end transaction tracing and traces for DF are correlated finally correctly. 
I found out that also deprecated SDK MS.Azure.ServiceBus may break the correlation...
But to make this feature work on bigger project is almost impossible.

**Comment by dlebee** [comment]
@JakubMajsniar using net6 with version 4 of azure function runtime and system.diagnosticsource v6.0.0 did the trick tor me

**Comment by JakubMajsniar** [comment]
@dlebee yes, for me too. Also in mentioned bigger project (I was able to upgraded all necessary stuff) and it is working. But with V3 version of runtime the latest working version of DiagnosticSource is 4.7.0. There are 474 packages which are dependent on DiagnosticSource and newer versions of them may need and bring newer version of DiagnosticSource. We have also found solution in our project where we had just downgrade 3 very common packages dependent on DiagnosticSource to use 4.7.0 version of it and project could stay in .netcore 3.1 and V3 runtime. 
I just wanted to share our experience :)
